File: backend/src/main/java/com/autocare/backend/job/entity/Job.java
package com.autocare.backend.job.entity;

import com.autocare.backend.auth.entity.User;
import com.autocare.backend.job.entity.enums.JobStage;
import com.autocare.backend.job.entity.enums.JobStatus;
import com.autocare.backend.job.entity.enums.JobType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

@Entity
@Table(name = "jobs")
@Getter
@Setter
@NoArgsConstructor
public class Job {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "id", updatable = false, nullable = false)
    private UUID id;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 50)
    private JobType type;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    private JobStatus status = JobStatus.PENDING;

    @Enumerated(EnumType.STRING)
    @Column(name = "current_stage", length = 100)
    private JobStage currentStage = JobStage.PENDING;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "payload", columnDefinition = "jsonb")
    private Map<String, Object> payload;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "result", columnDefinition = "jsonb")
    private Map<String, Object> result;

    @Column(name = "error_message", columnDefinition = "text")
    private String errorMessage;

    @Column(name = "retry_count", nullable = false)
    private Integer retryCount = 0;

    @Column(name = "max_retries", nullable = false)
    private Integer maxRetries = 3;

    @Column(name = "client_request_id", unique = true)
    private UUID clientRequestId;

    @Column(name = "last_heartbeat_at")
    private Instant lastHeartbeatAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt = Instant.now();

    @Column(name = "started_at")
    private Instant startedAt;

    @Column(name = "completed_at")
    private Instant completedAt;

    @Column(name = "expires_at")
    private Instant expiresAt;

    @NotNull
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;
}

File: mobile/src/types/job.ts
export type JobType = 'DIGITAL_TWIN_GENERATION' | 'OCR' | 'DIAGNOSTICS' | 'IMAGE_ANALYSIS' | 'PDF_PARSING';

export type JobStatus = 'PENDING' | 'RUNNING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';

export type JobStage =
  | 'PENDING'
  | 'VALIDATING_REQUEST'
  | 'LOOKING_FOR_TEMPLATE'
  | 'GENERATING_TEMPLATE'
  | 'SAVING_TEMPLATE'
  | 'CLONING_DIGITAL_TWIN'
  | 'FINALIZING'
  | 'COMPLETED'
  | 'FAILED'
  | 'CANCELLED';

export interface JobResult {
  vehicleId?: string;
  [key: string]: any;
}

export interface Job {
  id: string;
  userId: string;
  type: JobType;
  status: JobStatus;
  currentStage: JobStage;
  payload?: Record<string, any>;
  result?: JobResult;
  errorMessage?: string;
  retryCount: number;
  createdAt: string;
  startedAt?: string;
  completedAt?: string;
  lastHeartbeatAt?: string;
}

export type JobResponse = Job;
